### DISCUSSION TITLE
Lock-free or wait-free

### DISCUSSION THREAD

**Discussion by DaveLanday** [spacejam/rust-crdt#148] [Q&A]
Are CRDTs Lock-free and Wait-free since their internal semantics handle conflict resolution? Are there examples of using a CRDT in multi-threaded environments?

**Comment by davidrusu** [comment]
I think the concurrency paradigm would be "Message Passing" as each thread is expected to have their own copy of the CRDT state and no thread should mutate another threads CRDT state.

**Reply by DaveLanday** [reply]
Oh, I see.

If I wanted to pack multiple threads in the same process, and all threads were updating the same count, could they update an atomic that some main thread uses to update the CRDT and then that CRDT is shared between processes (assume atomic is for counting)? I suppose it could still be what you say; every thread has a copy of the CRDT counter, and then sends it's op over a channel or something. If the counter needed to be merged with a remote machine, then you would serialize the op and apply on the other side.

**Reply by davidrusu** [reply]
The second option is closer to how CRDT's want to be used. Keep a copy of each CRDT counter in each thread, synchronize threads/procs by sending ops to each other.

So in a sense, you treat threads, procs, nodes when synching
